### ISSUE TITLE
Data Capture

### ISSUE THREAD

**Issue by Omniflux** [dgrnbrg/fuji-debug-protocol#3]
[fuji-capture-1682291573-notes.txt](https://github.com/dgrnbrg/fuji-debug-protocol/files/11306502/fuji-capture-1682291573-notes.txt)
[fuji-capture-1682291573-annotated.log](https://github.com/dgrnbrg/fuji-debug-protocol/files/11306503/fuji-capture-1682291573-annotated.log)
[fuji-capture-1682291573.log](https://github.com/dgrnbrg/fuji-debug-protocol/files/11306504/fuji-capture-1682291573.log)

Data capture attached, defrost mode indicator on remote turned on at some point, I think near the end, and at a quick glance it appears one of the fan states may be missing, so may need to redo some or all.

**Comment by dgrnbrg** [comment]
This annotated log is along the lines of what will be ideal, but it would be really helpful if it included the lines like `[17:34:38][D][FujiHeatPump:289]:  mSrc: 32 mDst: 1 mType: 0 write: 0 login: 0 unknown: 1 onOff: 1 temp: 24, mode: 5 cP:1 uM:0 cTemp:24 acError:0 ` as well, since that's our closest to an already decomposed protocol.

There are 2 specific focus areas we should start with: the init protocol, and whether there's a checksum in the protocol.

For the init protocol, let's see, for the 2-3 devices (primary, A/C, [and secondary controller]), what the src and dst, mType, login, and unknown bits are for the entire exchange. I suspect we have a bug in our firmware's login/register handling.

For the checksum, let's investigate whether or why magic changes, and are there any other data that change without us changing something.

I can see that addresses 0, 1, and 32 are in use. I believe that 32 is the controller, and 0/1 are broadcast & A/C (I don't recall which is which off the top of my head.)

I should be able to really look into this in a couple of weeks.

**Comment by Omniflux** [comment]
Yes, it looks like there are several issues. I have switched from capturing logs with ESPHome to using Wireshark to examine the protocol - see [fujitsu-airstream-dissector](https://github.com/Omniflux/fujitsu-airstream-dissector).

I have picked up a UTY-RSNUM (simple remote) and UTY-TFNXZ2 (wifi) in addition to the UTY-RNNUM (standard remote) I already had.

It is generally accepted that only two remotes can be used simultaneously, but I suspect this is a limit of the remotes available, and not of the protocol. It appears the indoor unit looks for a primary remote, the primary remote looks for a secondary remote, and the secondary remote looks for a tertiary remote, despite Fujitsu not documenting any remotes which can have their address set as such.

Address fields may not be 7 bits, there may be more flags there with separate address space for indoor units and remotes...or not.

The RSNUM manual shows it can be connected to multiple indoor units simultaneously with the indoor unit address set via DIP switch. My indoor unit manuals do not show how to do this, so may not support it, and if so I will be unable to test.

I may be missing something, but I think the login and broadcast bits in the preexisting code may be completely wrong.

The indoor unit I am testing with does not support the swing options, and when I press the swing button on the remote, the remote shows it is not available without transmitting anything to the indoor unit. The indoor unit probably reports these options are unavailable in the login frame. May need to emulate an indoor unit to flip these bits individually and see what changes on the remote.

A checksum is looking unlikely.

**Comment by Omniflux** [comment]
I think I've extracted everything I can from the protocol that I can with the equipment I have - see project linked above.

I've confirmed addresses are only 4 bits + 1 type field. Does not appear any login occurs, "login" packet is actually optional features/capabilities request. No broadcast bit as all packets are broadcast, no destination address it is actually token handoff. No checksum.

The reliability issue may be a simple bug in the transmit code. There is no check that we have the token and a 100ms delay before TX. We should TX immediately when we have the token and not at all if we do not. Otherwise we cause a collision on the bus.
